src/Universum/Container/Utils.hs

```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE Trustworthy #-}
+
+-- | Utility functions on top of 'Container' typeclass.
+
+module Universum.Container.Utils
+       ( groupByFst
+       , groupByKey
+       , groupByKeyBy
+
+       , groupOn
+       ) where
+
+import Data.Function (id, on, (.))
+import Data.List.NonEmpty (NonEmpty (..), groupBy)
+
+import Universum.Base
+import Universum.Container.Class
+
+-- $setup
+-- >>> import Data.Function
+-- >>> import GHC.Num
+-- >>> import Data.Text (toLower)
+
+-- | Variation of 'groupByKey' that accepts the comparison function on keys
+-- explicitly. This function must define a (computable) equivalence relation.
+--
+-- Among multiple keys appearing in the same group it will pick the leftmost
+-- one as the representer of the equivalence class.
+--
+-- >>> groupByKeyBy (<=) id [(1, 10), (2, 11), (0, 12), (1, 13), (3, 14)]
+-- [(1,10 :| [11]),(0,12 :| [13,14])]
+groupByKeyBy
+  :: Container t
+  => (k -> k -> Bool) -> (Element t -> (k, v)) -> t -> [(k, NonEmpty v)]
+groupByKeyBy kcmp split = start . toList
+  where
+    start [] = []
+    start (a : as)
+      | (k, v) <- split a
+      , let (ys, zs) = go k as
+      = (k, v :| ys) : zs
+
+    go _ [] = ([], [])
+    go ko (a : as)
+      | (kn, v) <- split a
+      = if ko `kcmp` kn
+         then let (vs, ws) = go ko as
+                 in (v : vs, ws)
+         else let (vs, ws) = go kn as
+                 in ([], (kn, v :| vs) : ws)
+
+-- | Variation of 'groupByKey' that matches mapped keys on equality.
+--
+-- >>> groupByKeyOn toLower id [("A", 1), ("a", 2), ("b", 3)]
+-- [("A",1 :| [2]),("b",3 :| [])]
```

Comment:
This is a good example, but there should also be one or more with non-`id` projection (probably not the right word) functions.